src/main/java/net/openhft/chronicle/wire/AbstractWire.java

```diff
@@ -438,7 +438,7 @@ public boolean writeEndOfWire(long timeout, TimeUnit timeUnit, long lastPosition
                     pos += BytesUtil.padOffset(pos);
 
                 if (MEMORY.safeAlignedInt(pos)) {
-                    if (bytes.compareAndSwapInt(pos, 0, END_OF_DATA)) {
+                    if (bytes.readVolatileInt(pos) == 0 && bytes.compareAndSwapInt(pos, 0, END_OF_DATA)) {
```

Comment:
This works on the assumption that once `END_OF_DATA` is written, it isn't changed to something else. If it is, the `readVolatileInt` might evaluate to false, where a subsequent `compareAndSwap` might have resulted in a change.